tracing-serde/src/lib.rs

```diff
@@ -186,152 +188,214 @@ use tracing_core::{
     span::{Attributes, Id, Record},
 };
 
-pub mod fields;
+#[cfg(not(feature = "std"))]
+type TracingVec<T> = heapless::Vec<T, 32>;
 
-#[derive(Debug)]
-pub struct SerializeField(Field);
+#[cfg(not(feature = "std"))]
+type TracingMap<K, V> = heapless::FnvIndexMap<K, V, 32>;
 
-impl Serialize for SerializeField {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-    where
-        S: Serializer,
-    {
-        serializer.serialize_str(self.0.name())
-    }
-}
+#[cfg(feature = "std")]
+type TracingVec<T> = std::vec::Vec<T>;
 
-#[derive(Debug)]
-pub struct SerializeFieldSet<'a>(&'a FieldSet);
+#[cfg(feature = "std")]
+type TracingMap<K, V> = std::collections::HashMap<K, V>;
```

Comment:
this is kind of unfortunate, but changing a publicly-visible type based on a feature flag feels kind of like a violation of the "feature flags should be additive" guideline. 

in particular, the `From<TracingVec<...>>` and similar impls are concerning, because if you have some crate in a dependency tree that depends on `tracing-serde` with `default-features = false`, you can call the `From` impl with a `heapless::Vec`...but if some other crate in the dependency tree enables the `std` feature flag, that `From` impl now expects a *different* type and everything breaks.

I think we may have to figure out some solution where `TracingVec` and `TracingMap` are opaque newtypes which expose the same APIs on all feature combinations, and then have conversion methods from foreign types which can be feature-flagged. that way, enabling the `std` feature doesn't break any previously compiling code.

i think this means that the `heapless` dependency _may_ have to be non-optional, which...isn't *ideal*, but i can live with it.

Reply:
Hmm, I'm not against this idea, and I see what you're saying.

I think "make it a newtype, and impl From hvec always, and svec if `std` is active" isn't a bad plan. I can give it a try tomorrow.

Again, since we always have a known size of `32`, I *could* impl the vec/map types locally, but I wouldn't be super happy about it (heapless is much more battle tested, I'd probably just end up copy/pasting the impl from there).